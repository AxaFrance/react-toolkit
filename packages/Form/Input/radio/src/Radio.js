import React from 'react';
import PropTypes from 'prop-types';
import {
  InputList,
  InputConstants as Constants,
  withInput,
  omit,
} from '@axa-fr/react-toolkit-form-core';
import RadioItem from './RadioItem';
import RadioModes from './RadioModes';

const omitProperties = omit(['mode', 'helpMessage', 'id']);

const Radio = ({
  isVisible,
  className,
  disabled,
  options,
  value,
  name,
  onBlur,
  onFocus,
  readOnly,
  classModifier,
  children,
  onChange,
  ...otherProps
}) =>
  options.map(option => {
    const isChecked = option.value === value;
    return (
      <RadioItem
        key={option.value}
        id={option.id}
        value={option.value}
        label={option.label}
        isChecked={isChecked}
        name={name}
        onChange={onChange}
        onBlur={onBlur}
        onFocus={onFocus}
        readOnly={readOnly}
        isVisible={isVisible}
        disabled={InputList.isDisabled(option, disabled)}
        className={className}
        classModifier={option.classModifier}
        {...omitProperties(otherProps)}>
        {children}
      </RadioItem>
    );
  });

const propTypes = {
  ...Constants.propTypes,
  options: PropTypes.array.isRequired,
  value: PropTypes.string,
  mode: PropTypes.oneOf([
    RadioModes.default,
    RadioModes.classic,
    RadioModes.inline,
  ]),
};
const defaultClassName = '';
const defaultProps = {
  ...Constants.defaultProps,
  value: '',
  classModifier: defaultClassName,
  mode: RadioModes.default,
};

const handlers = {
  onChange: ({ onChange, name, id }) => e => {
    onChange({
      value: e.value,
      name,
      id,
    });
  },
};

const props = ({ mode }) => {
  let className = '';
  switch (mode) {
    case RadioModes.classic:
      className = 'af-form__radio';
      break;
    case RadioModes.inline:
      className = 'af-form__radio-inline';
      break;
    default:
      className = 'af-form__radio-custom';
      break;
  }
  return { className };
};

const EnhancedComponent = withInput(
  defaultClassName,
  propTypes,
  defaultProps,
  handlers,
  props
)(Radio);

EnhancedComponent.propTypes = propTypes;
EnhancedComponent.defaultProps = defaultProps;

EnhancedComponent.Clone = InputList.Clone;
EnhancedComponent.displayName = Radio.name;

export default EnhancedComponent;
