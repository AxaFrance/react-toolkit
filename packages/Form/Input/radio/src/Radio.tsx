import React, { ComponentProps } from 'react';
import RadioItem from './RadioItem';
import RadioModes from './RadioModes';

type Option = import('@axa-fr/react-toolkit-core/dist/InputManager').Option;
interface RadioProps
  extends Omit<ComponentProps<typeof RadioItem>, 'id' | 'label'> {
  className?: string;
  classModifier?: string;
  options: Option[];
  mode?: string;
}

const Radio: React.FC<RadioProps> = ({
  className,
  classModifier,
  options,
  value = '',
  name,
  mode = RadioModes.default,
  children,
  disabled,
  ...otherProps
}) => {
  const classNameWithMode = getClassName(mode);
  const optionItems = options.map((option: Option) => {
    const isChecked = option.value === value;
    return (
      <RadioItem
        key={option.value}
        id={option.id}
        value={option.value}
        label={option.label}
        checked={isChecked}
        disabled={option.disabled || disabled}
        className={classNameWithMode}
        classModifier={classModifier}
        optionClassName={className}
        {...otherProps}>
        {children}
      </RadioItem>
    );
  });
  return <>{optionItems}</>;
};

const getClassName = (mode: string): string => {
  switch (mode) {
    case RadioModes.classic:
      return 'af-form__radio';
    case RadioModes.inline:
      return 'af-form__radio-inline';
    default:
      return 'af-form__radio-custom';
  }
};

export default Radio;
