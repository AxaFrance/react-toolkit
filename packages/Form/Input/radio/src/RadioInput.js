import React from 'react';
import PropTypes from 'prop-types';

import {
  Field,
  FieldConstants as Constants,
} from '@axa-fr/react-toolkit-form-core';
import { InputManager } from '@axa-fr/react-toolkit-core';

import Radio from './Radio';
import RadioModes from './RadioModes';

const propTypes = {
  ...Constants.propTypes,
  options: PropTypes.array.isRequired,
  value: PropTypes.string,
  mode: PropTypes.oneOf([
    RadioModes.default,
    RadioModes.classic,
    RadioModes.inline,
  ]),
};

const defaultProps = {
  ...Constants.defaultProps,
  value: '',
  mode: RadioModes.default,
};

const RadioInput = props => {
  const {
    mode,
    messageType,
    message,
    className,
    classModifier,
    id,
    name,
    isVisible,
    options,
    classNameContainerLabel,
    classNameContainerInput,
    label,
    forceDisplayMessage,
    value,
    onChange,
    readOnly,
    disabled,
    children,
    ...otherProps
  } = props;
  if (!isVisible) {
    return null;
  }
  let rowModifier = classModifier;
  if (mode === RadioModes.classic) {
    rowModifier += ' label-top';
  }
  const newOptions = InputManager.getOptionsWithId(options);
  const firstId = InputManager.getFirstId(newOptions);
  return (
    <Field
      label={label}
      name={name}
      id={firstId}
      message={message}
      messageType={messageType}
      isVisible={isVisible}
      forceDisplayMessage={forceDisplayMessage}
      className={className}
      classModifier={rowModifier}
      classNameContainerLabel={classNameContainerLabel}
      classNameContainerInput={classNameContainerInput}>
      <Radio
        name={name}
        id={id}
        options={newOptions}
        value={value}
        mode={mode}
        onChange={onChange}
        readOnly={readOnly}
        disabled={disabled}
        classModifier={classModifier}
        {...otherProps}
      />
      {children}
    </Field>
  );
};

RadioInput.propTypes = propTypes;
RadioInput.defaultProps = defaultProps;

export default RadioInput;
