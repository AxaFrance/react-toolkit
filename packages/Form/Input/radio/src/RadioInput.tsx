import React, { ComponentProps } from 'react';
import { Field } from '@axa-fr/react-toolkit-form-core';
import { InputManager } from '@axa-fr/react-toolkit-core';
import Radio from './Radio';

import RadioModes from './RadioModes';

type RadioInputProps = ComponentProps<typeof Field> &
  ComponentProps<typeof Radio>;

const RadioInput: React.FC<RadioInputProps> = ({
  mode = RadioModes.default,
  messageType,
  message,
  className,
  classModifier,
  isVisible,
  options,
  classNameContainerLabel,
  classNameContainerInput,
  label,
  forceDisplayMessage,
  children,
  ...radioProps
}) => {
  const fieldClassModifier = `${classModifier ?? ''}${
    classModifier && mode === RadioModes.classic ? ' ' : ''
  }${mode === RadioModes.classic ? 'label-top' : ''}`;
  const newOptions = InputManager.getOptionsWithId(options);
  const firstId = InputManager.getFirstId(newOptions);
  const radioClassModifier = `${classModifier ?? ''}${
    classModifier && forceDisplayMessage && messageType ? ' ' : ''
  }${forceDisplayMessage && messageType ? `${messageType}` : ''}`;
  return (
    <Field
      label={label}
      id={firstId}
      message={message}
      messageType={messageType}
      isVisible={isVisible}
      forceDisplayMessage={forceDisplayMessage}
      className={className}
      classModifier={fieldClassModifier}
      classNameContainerLabel={classNameContainerLabel}
      classNameContainerInput={classNameContainerInput}>
      <Radio
        options={newOptions}
        mode={mode}
        classModifier={radioClassModifier}
        className={className}
        {...radioProps}
      />
      {children}
    </Field>
  );
};

export default RadioInput;
