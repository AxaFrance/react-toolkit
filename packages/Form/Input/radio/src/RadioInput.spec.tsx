import React, { ComponentProps } from 'react';
import { shallow } from 'enzyme';
import { Field, MessageTypes } from '@axa-fr/react-toolkit-form-core';
import RadioInput from './RadioInput';
import Radio from './Radio';
import { RadioModes } from '.';

describe('<RadioInput />', () => {
  let radioInputProps: ComponentProps<typeof RadioInput>;

  beforeEach(() => {
    radioInputProps = {
      message: 'test',
      messageType: MessageTypes.error,
      options: [
        { label: 'Option 1', value: '1', id: 'option1_id' },
        { label: 'Option 2', value: '2', id: 'option2_id' },
      ],
      id: '123',
      label: 'test',
    };
  });

  const Act = () =>
    shallow<typeof RadioInput>(<RadioInput {...radioInputProps} />);

  it('should have correct html rendered', () => {
    const result = Act();
    expect(result.html).toMatchSnapshot();
  });

  describe('Radio Child', () => {
    it.each`
      classModifier | expected
      ${''}         | ${'error'}
      ${'test'}     | ${'test error'}
      ${undefined}  | ${'error'}
      ${null}       | ${'error'}
    `(
      'should pass "$expected" to classModifier when classModifier is $classModifier',
      ({ classModifier, expected }) => {
        radioInputProps.forceDisplayMessage = true;
        radioInputProps.classModifier = classModifier;
        const result = Act();
        expect(result.find(Radio).props().classModifier).toEqual(expected);
      }
    );

    it.each`
      messageType
      ${null}
      ${undefined}
      ${''}
    `(
      'should pass "test" to classModifier when messageType is $messageType',
      ({ messageType }) => {
        radioInputProps.classModifier = 'test';
        radioInputProps.messageType = messageType;
        radioInputProps.forceDisplayMessage = true;
        const result = Act();
        expect(result.find(Radio).props().classModifier).toEqual('test');
      }
    );

    it.each`
      forceDisplayMessage | expected
      ${true}             | ${'error'}
      ${false}            | ${''}
      ${undefined}        | ${''}
      ${null}             | ${''}
    `(
      'should pass "$expected" to classModifier when forceDisplayMessage is $forceDisplayMessage',
      ({ forceDisplayMessage, expected }) => {
        radioInputProps.forceDisplayMessage = forceDisplayMessage;
        const result = Act();
        expect(result.find(Radio).props().classModifier).toEqual(expected);
      }
    );
  });

  describe('Field Child', () => {
    it.each`
      classModifier | expected
      ${''}         | ${'label-top'}
      ${'test'}     | ${'test label-top'}
      ${undefined}  | ${'label-top'}
      ${null}       | ${'label-top'}
    `(
      'should pass "$expected" to classModifier when classModifier is $classModifier',
      ({ classModifier, expected }) => {
        radioInputProps.mode = RadioModes.classic;
        radioInputProps.classModifier = classModifier;
        const result = Act();
        expect(result.find(Field).props().classModifier).toEqual(expected);
      }
    );

    it('should pass "test" to classModifier when mode is default', () => {
      radioInputProps.mode = RadioModes.default;
      radioInputProps.classModifier = 'test';
      const result = Act();
      expect(result.find(Field).props().classModifier).toEqual('test');
    });

    it.each`
      mode                  | expected
      ${RadioModes.classic} | ${'label-top'}
      ${RadioModes.default} | ${''}
      ${''}                 | ${''}
      ${undefined}          | ${''}
      ${null}               | ${''}
    `(
      'should pass "$expected" to classModifier when mode is $mode',
      ({ mode, expected }) => {
        radioInputProps.mode = mode;
        const result = Act();
        expect(result.find(Field).props().classModifier).toEqual(expected);
      }
    );
  });
});
