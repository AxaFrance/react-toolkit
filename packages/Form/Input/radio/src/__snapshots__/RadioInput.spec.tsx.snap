// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioInput /> should have correct html rendered 1`] = `[Function]`;
